### DISCUSSION TITLE
issue using closed loop stepper

### DISCUSSION THREAD

**Discussion by EdFleta** [kachurovskiy/nanoels#86] [Q&A]
Hi Maxim,

With the new H2 version I am using a closed loop stepper from Cloudray (Nema 23 + CL23 driver). Since the minimum step configuration is 800 with the driver, I have modified ligne 7 of the software to : "#define MOTOR_STEPS 800.0".

However, in some cases (without being able to define precise circumstances - but more generally at first start) the stepper runs very slowly and I have to keep pushing the running buttons left or right for a while until the motor reaches full speed.

Is there any config that I should adapt for closed loop steppers?

best regards, eddie

**Comment by kachurovskiy** [comment]
Hi @EdFleta, I've tried using H2 with a closed loop stepper today for the first time to check this :) I'm having different problems - it sometimes makes a few steps back before moving forward. Sometimes it moves forward by too much :( I suspect that my cheap servo might be to blame... That said, I love how smooth the lead screw movements are 😻 

As for your problem, it would really help to have a video of it happening. E.g. just put your phone in recording mode and play with the device until the problem starts happening. You can also cut out the irrelevant parts of the video on your phone e.g. in Google Photos app before uploading it here. Cheers!

**Reply by EdFleta** [reply]
Thanks for your rapid answer Maxim, I'll try your recommendation. However it looks like a random issue and I am not sure I can capture it easily. 
I will also post some pictures of my set up with H2. It is now finalised.
Let me take this opportunity to congratulate you for the quality of your work.
best regards,

eddie

**Reply by Johannmupa** [reply]
Just today I have connected it with closed loop stepper. It seems to work nice and the motor runs smooth and quiet. Closed loop drivers have the microstep setting "default" written above the 800 steps. Usually this ist halfstep 400 steps per revolutio. I can not confirm it yet, because it came without any documetation other than printed on the device. 
But I have trouble with the 5 contacts at the AB input. As if the buttons were pressed randomly. I will check the 1 Mohm resistor to ground. 
Big trouble makes my VFD drive on the lathe. Not always but sometimes the Arduino sreen shows random characters running over the screen when I start the spindle.  I have to try a shielded motor cable. 
Spindle encoder is not mounted yet. I just got 2GT timing belt and wheels today. I hope the belt drive is more quiet than gears.

**Reply by kachurovskiy** [reply]
@Johannmupa interesting, please consider starting a new discussion with more info about your problem - ideally with photos/videos showing it. Thank you!

**Reply by Johannmupa** [reply]
Dear Maxim,
after fixing my EMI problems I observe the same problem with the closed loop stepper and h2. When I press the left or right button it sometime does not moove sometimes it moves backwards sometimes just the increment that i choose. The position counter is running even when the stepper did not. I can observe it very well because i have a graduated handwheel at the right side of the leadscrew. It depends how long the button is pressed. When pressed nearly as long as it takes to repeat the steps (0.1 0.01 or 0.001) it works perfect. The steps that were lost this way add up when switch to the ON mode. The stop positions are different then.

**Reply by kachurovskiy** [reply]
Hi @Johannmupa, I've changed the code a bit and now there's `DISABLE_STEPPER_WHEN_RESTING ` parameter in https://github.com/kachurovskiy/nanoels/blob/main/h2/h2.ino which is false by default. I used to disable stepper when it's not moving but it turned out not useful for closed loop steppers for 2 reasons:

- closed loop stepper drivers, unlike open loop drivers, have a delay when enabling - they ignore some impulses for 100-300ms after they were enabled
- closed loop stepper drivers don't hum or heat up when idle so there's no benefit disabling them

If you don't want to download new software version just for this one fix, you can comment out line `digitalWrite(ENA, LOW);` in your code to never disable the stepper or disconnect the corresponding ENA wire on your stepper driver. There's a good chance it will help with your issue, at least partially. Cheers!

**Reply by Johannmupa** [reply]
Dear Maxim,
I have uploaded the latest code and it works fine. After a lot of testing there are no issues. 
thanks a lot

**Reply by kachurovskiy** [reply]
Awesome, congratulations 🥳

**Reply by EdFleta** [reply]
Hi, I did the same and I confirm closed loop steppers work better with this version. (especially on smaller increments)
eddie

**Reply by kachurovskiy** [reply]
@EdFleta consider taking a short video of your setup and problems and uploading it here. Cheers!

**Comment by Johannmupa** [comment]
Thanks, I've just posted it

Am Sa., 19. Nov. 2022 um 19:46 Uhr schrieb Maxim Kachurovskiy <
***@***.***>:
